php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,559 @@
+				'saveButton' => __( 'Save', 'customize-widgets-plus' ),
+				'cancelButton' => __( 'Cancel', 'customize-widgets-plus' ),
+				'shareButton' => __( 'Share URL to preview', 'customize-widgets-plus' ),
+				'updateMsg' => __( 'Clicking "Save" will update the current snapshot.', 'customize-widgets-plus' ),
+				'errorMsg' => __( 'The snapshot could not be saved.', 'customize-widgets-plus' ),
+				'previewTitle' => __( 'Preview URL', 'customize-widgets-plus' ),
+				'formTitle' => ( $this->snapshot->is_preview() ?
+					__( 'Update Snapshot', 'customize-widgets-plus' ) :
+					__( 'Snapshot Scope', 'customize-widgets-plus' )
+				),
+				'dirtyLabel' => __( 'Diff Snapshot (preview dirty settings)', 'customize-widgets-plus' ),
+				'fullLabel' => __( 'Full Snapshot (preview all settings)', 'customize-widgets-plus' ),
+			),
+		);
+
+		// Export data to JS.
+		wp_scripts()->add_data(
+			$this->plugin->script_handles[ $handle ],
+			'data',
+			sprintf( 'var _customizeWidgetsPlusCustomizeSnapshot = %s;', wp_json_encode( $exports ) )
+		);
+	}
+
+	/**
+	 * Get the Customize_Snapshot instance.
+	 *
+	 * @return Customize_Snapshot
+	 */
+	public function snapshot() {
+		return $this->snapshot;
+	}
+
+	/**
+	 * Save a snapshot.
+	 *
+	 * @param \WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 * @param string $status The post status.
+	 * @return null|\WP_Error
+	 */
+	public function save( \WP_Customize_Manager $manager, $status = 'draft' ) {
+		$new_setting_ids = array_diff( array_keys( $this->post_data ), array_keys( $manager->settings() ) );
+		$manager->add_dynamic_settings( $new_setting_ids );
+
+		if ( ! empty( $this->contextual_post_data ) ) {
+			foreach ( $this->contextual_post_data as $setting_id => $contextual ) {
+				$this->snapshot->set_contextual( $setting_id, $contextual['value'], $contextual['dirty'] );
+			}
+		}
+
+		foreach ( $manager->settings() as $setting ) {
+			if ( $this->can_preview( $setting, $this->post_data ) ) {
+				$post_data = $this->post_data[ $setting->id ];
+				$this->snapshot->set( $setting, $post_data['value'], $post_data['dirty'] );
+			}
+		}
+
+		$r = $this->snapshot->save( $status );
+		if ( is_wp_error( $r ) ) {
+			status_header( 500 );
+			wp_send_json_error( $r->get_error_message() );
+		}
+	}
+
+	/**
+	 * Save snapshots via AJAX.
+	 *
+	 * Fires at `customize_save_after` to update and publish the snapshot.
+	 *
+	 * @param \WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 */
+	public function save_snapshot( \WP_Customize_Manager $manager ) {
+		if ( ! current_user_can( 'customize' ) ) {
+			status_header( 403 );
+			wp_send_json_error( 'customize_not_allowed' );
+		} else if ( 'POST' !== $_SERVER['REQUEST_METHOD'] ) {
+			status_header( 405 );
+			wp_send_json_error( 'bad_method' );
+		}
+
+		$uuid = null;
+		if ( ! empty( $_POST['snapshot_uuid'] ) ) {
+			if ( empty( $this->post_data ) ) {
+				status_header( 400 );
+				wp_send_json_error( 'missing_snapshot_customized' );
```

Comment:
@valendesigns instead of short-circuiting the `customize-save` request with an error, how about amending the `customize_save_response` filter to include these errors. In this way, any errors happening for snapshots won't cause other plugins from working properly. The `customize_save_response` filter is how general messages can be sent back.